core/src/main/java/org/elasticsearch/action/admin/indices/shards/IndicesShardsStoresResponse.java

```diff
@@ -0,0 +1,362 @@
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentBuilderString;
+import org.elasticsearch.rest.RestStatus;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.ExceptionsHelper.detailedMessage;
+
+/**
+ * Consists of {@link ShardStoreStatus}s for request indices grouped by
+ * indices and shard ids and a list of node {@link Failure}s encountered
+ */
+public class IndicesShardsStoresResponse extends ActionResponse implements ToXContent {
+
+    /**
+     * Shard store information from a node
+     */
+    public static class ShardStoreStatus implements Streamable, ToXContent {
+        private DiscoveryNode node;
+        private long version;
+        private Throwable storeException;
+
+        private ShardStoreStatus() {
+        }
+
+        ShardStoreStatus(DiscoveryNode node, long version, Throwable storeException) {
+            this.node = node;
+            this.version = version;
+            this.storeException = storeException;
+        }
+
+        /**
+         * node the shard belongs to
+         */
+        public DiscoveryNode getNode() {
+            return node;
+        }
+
+        /**
+         * version of the shard, the highest
+         * version of the replica shards are
+         * promoted to primary if needed
+         */
+        public long getVersion() {
+            return version;
+        }
+
+        /**
+         * exception while trying to open the
+         * shard index or from when the shard failed
+         */
+        public Throwable getStoreException() {
+            return storeException;
+        }
+
+        private static ShardStoreStatus readShardStoreStatus(StreamInput in) throws IOException {
+            ShardStoreStatus shardStoreStatus = new ShardStoreStatus();
+            shardStoreStatus.readFrom(in);
+            return shardStoreStatus;
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            node = DiscoveryNode.readNode(in);
+            version = in.readLong();
+            if (in.readBoolean()) {
+                storeException = in.readThrowable();
+            }
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            node.writeTo(out);
+            out.writeLong(version);
+            if (storeException != null) {
+                out.writeBoolean(true);
+                out.writeThrowable(storeException);
+            } else {
+                out.writeBoolean(false);
+            }
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject(Fields.NODE);
```

Comment:
can we take the code from org/elasticsearch/cluster/ClusterState.java:400 (serializing DiscoNodes to xcontent) and put in on DiscoveryNode and reuse it here? (note the tricky node id code)